libpam/pam_modutil_sanitize.c

```diff
@@ -128,15 +132,30 @@ close_fds(void)
 	int fd;
 	struct rlimit rlim;
 
-	if (getrlimit(RLIMIT_NOFILE, &rlim) || rlim.rlim_max > MAX_FD_NO)
-		fd = MAX_FD_NO;
-	else if (rlim.rlim_max < MIN_FD_NO)
-		fd = MIN_FD_NO;
-	else
-		fd = rlim.rlim_max - 1;
-
-	for (; fd > STDERR_FILENO; --fd)
-		close(fd);
+	/* 
+	 * if /proc is mounted, we can optimize what fd can be closed,
+ 	 * but if it isn't available, fall back to the previous behavior.
+	 */
+	if ((dir = opendir("/proc/self/fd")) != NULL) {
```

Comment:
what if /proc/self/fd is not a valid proc filesystem?

Reply:
I'm confused. Several parts of pam code relies on /proc, and I can't see any check if /proc is a valid proc fs. I'm not against a change on that, but please explain me how to check if /proc is a valid fs? Using functions like getmntent (which can check a given mnt_dir and mnt_type) reads /etc/mtab, which is itself a symlink to /proc/self/mounts...

Reply:
I would not really care about this. So severally bogus setup is not something we want to support anyway. /proc must be procfs or anything can happen.

Reply:
I don't want any regressions in case of /proc is not mounted, so I suggest adding a statfs(dirfd(dir)) and checking that f_type matches PROC_SUPER_MAGIC.

Ideally, we could use close_range syscall when linux finally implements it, but we are not there yet.

Reply:
The check for PROC_SUPER_MAGIC would require AC_CHECK_HEADERS(linux/magic.h),
#include <linux/magic.h> under #ifdef HAVE_LINUX_MAGIC_H,
and the whole check would have to be guarded with #ifdef PROC_SUPER_MAGIC.

Reply:
But why there would be any regression? If /proc is not mounted there won't be /proc/self/fd and so the old code will be used and if somebody with sufficient rights makes an effort to break his environment to create some bogus thing in /proc/self/fd without mounting /proc I do not know why he should not bear its consequences. He can as well do bind mount from some another procfs directory to /proc/self/fd and you would not be able to detect it even with this trick.

Reply:
I've just pushed a new commit implementing this check. Would you please review it and let me know if I have to squash my two commits into a single one.